### PR TITLE
Document speed comparison

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?
Extend documentation on how the Pl is running almost the same fast as vanilla PT
Follow-up from #1587

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edenlightning** [comment]
+ make sure to add a todo for memory parity to catch leaks

**Comment by Borda** [comment]
config with k80
```
pt_times: [2.3914196679834276, 2.394015436992049, 2.378540376899764, 2.3768294940236956, 2.3556928229518235, 2.3486300229560584, 2.3433066660072654, 2.3670845460146666, 2.3849029790144414, 2.3830047389492393, 2.3483149090316147, 2.344499905942939, 2.376186921959743, 2.372628123033792, 2.3533506059320644, 2.3700953329680488, 2.332227217964828, 2.360867984010838, 2.3292078389786184]
pl_times: [2.553553794976324, 2.550384434987791, 2.5761637160321698, 2.5912115490064025, 2.5808606880018488, 2.590446621994488, 2.5852647749707103, 2.5828406190266833, 2.587320258957334, 2.5930449219886214, 2.5799579489976168, 2.565676204045303, 2.58295069495216, 2.588471563067287, 2.575781085062772, 2.5723539519822225, 2.585605831933208, 2.576303143054247, 2.5885326160350814]
```

**Comment by edenlightning** [comment]
@Borda still relevant?

**Comment by Borda** [comment]
> @Borda still relevant?

yes, will do it next week...

File: docs/source/trainer.rst [hunk 1]
```diff
@@ -1718,3 +1718,14 @@ The metrics sent to the progress bar.
 
     progress_bar_metrics = trainer.progress_bar_metrics
     assert progress_bar_metrics['a_val'] == 2
+
+
+Benchmark with vanilla PyTorch
+------------------------------
+
+We have set regular benchmarking against PyTorch vanilla training loop on with RNN and simple MNIST classifier as per of out CI.
```
Comment (tchaton) [inline docs/source/trainer.rst:1726]: Can you add the date there ?

**Review by tchaton — APPROVED** [review]

File: docs/source/trainer.rst (hunk 1 above)
Comment (Borda) [reply docs/source/trainer.rst:1726]: any suggestion if put in the title or just as a footnote to the text?

**Comment by Borda** [comment]
image

File: benchmarks/test_basic_parity.py [hunk 2]
```diff
@@ -15,34 +30,33 @@
     (ParityModuleMNIST, 0.25),  # todo: lower this thr
 ])
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
-def test_pytorch_parity(tmpdir, cls_model, max_diff):
+def test_pytorch_parity(tmpdir, cls_model, max_diff, num_epochs: int = 4, num_runs: int = 3):
```
Comment (SkafteNicki) [inline benchmarks/test_basic_parity.py:33]:
```suggestion
def test_pytorch_parity(tmpdir, cls_model, max_diff: float, num_epochs: int = 4, num_runs: int = 3):
```

File: benchmarks/generate_comparison.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        if os.path.isfile(path_csv):
+            df_time = pd.read_csv(path_csv, index_col=0)
+        else:
+            vanilla = vanilla_loop(cls_model, num_epochs=NUM_EPOCHS, num_runs=NUM_RUNS)
+            lightning = lightning_loop(cls_model, num_epochs=NUM_EPOCHS, num_runs=NUM_RUNS)
+
+            df_time = pd.DataFrame({'vanilla PT': vanilla['durations'][1:], 'PT Lightning': lightning['durations'][1:]})
+            df_time /= NUM_RUNS
+            df_time.to_csv(os.path.join(PATH_HERE, f'dump-times_{cls_model.__name__}.csv'))
+        # todo: add also relative X-axis ticks to see both: relative and absolute time differences
```
Comment (Borda) [inline benchmarks/generate_comparison.py:43]: @tchaton

**Review by tchaton — APPROVED** [review]

File: docs/source/benchmarking.rst [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+In this section we set grounds for comparison between vanilla PyTorch and PT Lightning for most common scenarios.
+
+Time comparison
+---------------
+
+We have set regular benchmarking against PyTorch vanilla training loop on with RNN and simple MNIST classifier as per of out CI.
+In average for simple MNIST CNN classifier we are only about 0.06s slower per epoch, see detail chart bellow.
+
+.. figure:: _images/benchmarks/figure-parity-times.png
+   :alt: Speed parity to vanilla PT
```
Comment (Borda) [inline docs/source/benchmarking.rst:13]:
```suggestion
   :alt: Speed parity to vanilla PT, created on 2020-12-16
```

File: docs/source/trainer.rst (hunk 1 above)
Comment (Borda) [reply docs/source/trainer.rst]: added as fig captions

**Review by SeanNaren — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by awaelchli — COMMENTED** [review]
I have the feeling this does not fit into the "best practices" section. But there are not many other options. Maybe @edenlightning has a good idea? :)

**Comment by awaelchli** [comment]
maybe @SeanNaren wants to include his ddp benchmarking results here too? :)

**Comment by Borda** [comment]
> I have the feeling this does not fit into the "best practices" section. But there are not many other options. Maybe @edenlightning has a good idea? :)

I think we can move it anytime, I just need this as base for memory benchmark :]

> maybe @SeanNaren wants to include his ddp benchmarking results here too? :) I've seen some great charts somewhere

that would be great, so another PR?

**Pull request merged**